Cache ILM policy name on IndexMetadata

joegallo opened the pull request:
Closes #83582

One little detail is that the `LIFECYCLE_NAME` value is defined on [IndexMetadata](https://github.com/elastic/elasticsearch/pull/83603/files#diff-d9e713c7c9f70d3b97888b7ed019ba4469397004333b0d0c13fa98442beef047R822) but it's still almost entirely referenced by way of [LifecycleSettings](https://github.com/elastic/elasticsearch/pull/83603/files#diff-4e8f046e86e7ec85e51de358c9773903205bebc9b34afacaab5d7d479837a39cR20). We need to have the value available in IndexMetadata in order to pull the value from the settings and set it into `lifecyclePolicyName`, but besides that everything about this still seems more `LifecycleSettings`-related to me.

joegallo committed:
Use policyName more consistently as a variable name

joegallo committed:
Hoist the ILM lifecycle policy name to IndexMetadata

rather than repeatedly looking it up in the settings all the time

joegallo committed:
Mark as nullable, add a test

joegallo committed:
Avoid LIFECYCLE_NAME_SETTING.getKey()

in favor of just using LIFECYCLE_NAME

joegallo committed:
Use Setting.timeSetting

for symmetry with all the rest of the Setting.whateverSetting calls
here.

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/83603.yaml

original-brownbear commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:
```diff
@@ -35,7 +34,7 @@
     public static final String SNAPSHOT_INDEX_NAME = "index.store.snapshot.index_name";
 
-    public static final Setting<TimeValue> LIFECYCLE_POLL_INTERVAL_SETTING = timeSetting(
+    public static final Setting<TimeValue> LIFECYCLE_POLL_INTERVAL_SETTING = Setting.timeSetting(
```
This seems unrelated?

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -1642,6 +1656,7 @@ public IndexMetadata build() {
                 DiskThresholdDecider.SETTING_IGNORE_DISK_WATERMARKS.get(settings),
                 tierPreference,
                 ShardsLimitAllocationDecider.INDEX_TOTAL_SHARDS_PER_NODE_SETTING.get(settings),
+                settings.get(IndexMetadata.LIFECYCLE_NAME),
```
This seems mildly hacky, should we just move the full Setting instance in here so we can just use `LIFECYCLE_NAME_SETTING.get(settings)` the standard way?

joegallo replied:
This has the `null` semantics I want, but the standard way doesn't -- `SOME_SETTING.get(settings)` returns the default value for the setting if the setting is absent, and in this case that's an empty string (because you can't define the default value for a setting as being `null`).

Of course, that's not impossible to work around, we could do a `hasText` check or something like that.

original-brownbear replied:
Ah I see. This is mildly scary because we seem to have been mixing both patterns for this setting before and now we're only using one that has `null` in it.
But I guess we use `if (Strings.isNullOrEmpty(policyName) == false) {` or the return as the right hand side of an equals so it should be good.

original-brownbear reviewed (commented):
Thanks Joe! just one quick question.

joegallo committed:
Fix this condition

Prior to this PR, this condition was not correct, and would never have been triggered -- it used:

    LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexMetadata.getSettings())

to read the policy name, and the .get there would return the default
value for the setting (empty string) if the setting was not present,
*not* null. So this condition would always see a non-null policyName
and never actually did anything.

We have some tests that exercise this scenario, though, and they
expect the error message that we throw from
IndexLifecycleTransition.validateTransition -- by throwing with that
same error message here those tests pass and the shortcut condition
here actually works as expected.

joegallo committed:
Merge branch 'master' into hoist-lifecycle-name-setting

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -759,6 +765,10 @@ public int getShardsPerNodeLimit() {
         return tierPreference;
     }
 
+    public String getLifecyclePolicyName() {
```
Let's add `@Nullable` here

joegallo replied:
👍, https://github.com/elastic/elasticsearch/pull/83603/commits/508447c8efa1f77d8d0de4e3dac6054a9271b90b

original-brownbear reviewed (approved):
LGTM (maybe add the annotation though, makes everyone's life easier). Thanks Joe!

joegallo committed:
Add nullable, and some explanations

joegallo committed:
Merge branch 'master' into hoist-lifecycle-name-setting

joegallo committed:
More explicit policy name validation

It would have NPE-ed on null, and AFAICT it would never have been
called with an empty string (and we already check for spaces).

joegallo committed:
Invert this check

This class already uses hasText for this elsewhere, might as well use
it consistently (it's shorter and clearer this way).

joegallo committed:
Merge branch 'master' into hoist-lifecycle-name-setting

joegallo committed:
Merge branch 'master' into hoist-lifecycle-name-setting

joegallo commented:
@elasticmachine run elasticsearch-ci/bwc
